### DISCUSSION TITLE
Simulating Rigid Bodies with charged constituents

### DISCUSSION THREAD

**Discussion by amruthesht** [glotzerlab/hoomd-blue#1789] [Support] [answered]
Hello All,

I'm trying to run rigid body simulations with charged constituent particles. I need help with the following two points.

 1. I noticed that the `charges` key has been removed from the rigid constituent dictionary and moved to `create_bodies()`. Does this mean `integrator.rigid = rigid` wouldn't automatically assign charges to the constituents? Is there a way to continue simulating with charged constituents?
 2. I understand that a GSD file doesn't store information on the constituent particles of a rigid body. However, is there a way to output the configurations actively in a format that also takes care of the constituent particle positions (so one can say visualize the simulation as a movie)?

Thank you!
Amruthesh

**Comment by joaander** [comment, marked as answer]
Unlike the other keys in `body` (type, position, and orientation): `constrain.Rigid` has never automatically set charges on constituents on every time step. Moving the dictionary to `create_bodies` reflects the actual behavior. When you call `create_bodies`, it will set the charges on the constituent particles that once.

GSD does store information on the constituent particles (including the charge). When you load the system state from a GSD file containing constituent particles, there is no need to call `create_bodies` - it is a convenience method to place particles that you could also place by other means.

To restate: If you are migrating from a previous version of HOOMD-blue, then there is *no* change in the behavior of `constrain.Rigid`. The only difference is that previously you set charges in `body` (which were used only once by `create_bodies`) and now you pass the charges to `create_bodies`. If you previously set constituent particles by some other means, the only change is that you no longer need to set charges in `body` that will go unused.

**Reply by amruthesht** [reply]
Thank you for your response.

Just to clarify, once the charges are placed using `create_bodies`, they will not be overwritten by the MD integrator when it is called (through `integrator.rigid = rigid`)(as it does for positions, orientations and typeids; explained as an example [here](https://hoomd-blue.readthedocs.io/en/latest/tutorial/06-Modelling-Rigid-Bodies/01-Introduction-to-Rigid-Bodies.html)), right?

As for outputting rigid bodies in GSD -- Is any info about the rigid body stored in the GSD, i.e., when a simulation state is initialized from a GSD, is there any info on the central particle and its corresponding constituents, or are they all treated as independent entities?

Thank you again,
Amruthesh

**Reply by joaander** [reply]
> Just to clarify, once the charges are placed using `create_bodies`, they will not be overwritten by the MD integrator when it is called (through `integrator.rigid = rigid`)(as it does for positions, orientations and typeids; explained as an example [here](https://hoomd-blue.readthedocs.io/en/latest/tutorial/06-Modelling-Rigid-Bodies/01-Introduction-to-Rigid-Bodies.html)), right?

Yes, the [reference docs](https://hoomd-blue.readthedocs.io/en/v4.7.0/module-md-constrain.html#hoomd.md.constrain.Rigid) are very clear on this point as well.

> As for outputting rigid bodies in GSD -- Is any info about the rigid body stored in the GSD, i.e., when a simulation state is initialized from a GSD, is there any info on the central particle and its corresponding constituents, or are they all treated as independent entities?

GSD stores both central and constituent particles and associates them with the [body tag](https://gsd.readthedocs.io/en/v3.3.0/schema-hoomd.html#chunk-particles-body). GSD does *not* store the parameters you set in `rigid.body`.

**Reply by amruthesht** [reply]
Thank you very much for the clarification!
